File: curvefit.py

from __future__ import annotations
import numpy as np


def _chord_length_parameterize(points):
    u = [0.0]
    for i in range(1, len(points)):
        u.append(u[-1] + np.linalg.norm(points[i] - points[i - 1]))
    total = u[-1] if u[-1] > 1e-9 else 1.0
    return [ui / total for ui in u]


def _bezier(t, ctrl):
    mt = 1 - t
    return (mt ** 3) * ctrl[0] + 3 * (mt ** 2) * t * ctrl[1] + 3 * mt * (t ** 2) * ctrl[2] + (t ** 3) * ctrl[3]


def _bezier_deriv(t, ctrl):
    mt = 1 - t
    return 3 * (mt ** 2) * (ctrl[1] - ctrl[0]) + 6 * mt * t * (ctrl[2] - ctrl[1]) + 3 * (t ** 2) * (ctrl[3] - ctrl[2])


def _q(ctrl, t):
    return _bezier(t, ctrl)


def _reparameterize(points, u, ctrl):
    new_u = []
    for i, p in enumerate(points):
        new_u.append(_newton_raphson_root_find(ctrl, p, u[i]))
    return new_u


def _newton_raphson_root_find(ctrl, point, u):
    d = _q(ctrl, u) - point
    qp = _bezier_deriv(u, ctrl)
    qpp = 6 * (1 - u) * (ctrl[2] - 2 * ctrl[1] + ctrl[0]) + 6 * u * (ctrl[3] - 2 * ctrl[2] + ctrl[1])
    denom = np.dot(qp, qp) + np.dot(d, qpp)
    if abs(denom) < 1e-9:
        return u
    return u - np.dot(d, qp) / denom


def _generate_bezier(points, u, t_hat1, t_hat2):
    n = len(points)
    A = np.zeros((n, 2, 2))
    for i, t in enumerate(u):
        A[i][0] = t_hat1 * 3 * (1 - t) ** 2 * t
        A[i][1] = t_hat2 * 3 * (1 - t) * t ** 2

    C = np.zeros((2, 2))
    Xv = np.zeros(2)
    p0, p3 = points[0], points[-1]
    for i, t in enumerate(u):
        C[0][0] += np.dot(A[i][0], A[i][0])
        C[0][1] += np.dot(A[i][0], A[i][1])
        C[1][0] = C[0][1]
        C[1][1] += np.dot(A[i][1], A[i][1])
        base = _bezier(t, (p0, p0, p3, p3)) - points[i]
        Xv[0] -= np.dot(A[i][0], base)
        Xv[1] -= np.dot(A[i][1], base)

    det_C0C1 = C[0][0] * C[1][1] - C[1][0] * C[0][1]
    alpha_l = alpha_r = 0.0
    if abs(det_C0C1) > 1e-12:
        det_C0X = C[0][0] * Xv[1] - C[1][0] * Xv[0]
        det_XC1 = Xv[0] * C[1][1] - Xv[1] * C[0][1]
        alpha_l = det_XC1 / det_C0C1
        alpha_r = det_C0X / det_C0C1

    # seg_len (straight p0->p3 distance) degenerates to ~0 on a CLOSED curve's
    # very first fit attempt, since points[0] and points[-1] are (nearly) the
    # same point even though the arc between them is the whole loop. When
    # that happens the classic Schneider "too small -> fall back to seg_len/3"
    # guard never fires (eps collapses to ~0 too), so an ill-conditioned
    # linear system (near-singular det_C0C1) can hand back an enormous or
    # negative alpha and the control points shoot off far outside the
    # drawing's bounding box. Use the polyline's actual arc length as a
    # size reference instead, and clamp alpha on BOTH sides (too small AND
    # too large), not just the too-small side.
    seg_len = np.linalg.norm(p0 - p3)
    arc_len = 0.0
    for i in range(1, len(points)):
        arc_len += np.linalg.norm(points[i] - points[i - 1])
    # Bound alpha by the SPAN of this point subset (its bounding-box
    # diagonal), not by arc_len: on a closed loop's first fit attempt
    # arc_len is the whole circumference while the chord p0->p3 is ~0, and
    # neither of those is the right yardstick for "how far a control point
    # may reasonably sit from its endpoint" -- a control point should never
    # need to reach further than the region the points themselves occupy.
    diag = float(np.hypot(points[:, 0].max() - points[:, 0].min(),
                           points[:, 1].max() - points[:, 1].min()))
    ref_len = seg_len if seg_len > 1e-6 else (arc_len / 3.0 if arc_len > 1e-6 else 1.0)
    fallback = ref_len / 3.0
    max_alpha = max(diag, ref_len, 1e-6)
    eps = 1e-6 * ref_len if ref_len > 0 else 1e-6
    if alpha_l < eps or alpha_r < eps or alpha_l > max_alpha or alpha_r > max_alpha:
        alpha_l = alpha_r = fallback

    c1 = p0 + t_hat1 * alpha_l
    c2 = p3 + t_hat2 * alpha_r
    return (p0, c1, c2, p3)


def _compute_max_error(points, u, ctrl):
    max_dist = 0.0
    split_point = len(points) // 2
    for i, p in enumerate(points):
        d = _q(ctrl, u[i]) - p
        dist = np.dot(d, d)
        if dist > max_dist:
            max_dist = dist
            split_point = i
    return max_dist, split_point


def _fit_cubic(points, t_hat1, t_hat2, error, out):
    if len(points) == 2:
        dist = np.linalg.norm(points[0] - points[1]) / 3.0
        p0, p3 = points[0], points[1]
        out.append((p0, p0 + t_hat1 * dist, p3 + t_hat2 * dist, p3))
        return

    u = _chord_length_parameterize(points)
    ctrl = _generate_bezier(points, u, t_hat1, t_hat2)
    max_err, split_point = _compute_max_error(points, u, ctrl)
    if max_err < error:
        out.append(ctrl)
        return

    for _ in range(4):
        u = _reparameterize(points, u, ctrl)
        ctrl = _generate_bezier(points, u, t_hat1, t_hat2)
        max_err, split_point = _compute_max_error(points, u, ctrl)
        if max_err < error:
            out.append(ctrl)
            return

    if split_point < 2:
        split_point = 2
    if split_point > len(points) - 2:
        split_point = len(points) - 2

    t_hat_center = points[split_point - 1] - points[split_point + 1]
    norm = np.linalg.norm(t_hat_center)
    t_hat_center = t_hat_center / norm if norm > 1e-9 else np.array([0.0, 0.0])

    _fit_cubic(points[: split_point + 1], t_hat1, t_hat_center, error, out)
    _fit_cubic(points[split_point:], -t_hat_center, t_hat2, error, out)


def _unit_tangent(a, b):
    v = b - a
    n = np.linalg.norm(v)
    return v / n if n > 1e-9 else np.array([0.0, 0.0])


def adaptive_max_error(points_xy, ratio: float = 0.012, lo: float = 1.2, hi: float = 22.0) -> float:
    """
    "Smooth finish" support: a fixed pixel max_error (e.g. 3.0) means a big
    drawing gets fit with lots of short segments (looks hand-drawn / jittery)
    while a tiny drawing gets over-simplified. Scaling the tolerance to the
    stroke's own size gives a consistent, size-independent result: fewer,
    longer, cleaner curve segments -- closer to what a vector-tracing tool
    (Illustrator "Image Trace", potrace...) would produce -- regardless of
    whether the design is drawn small or large on the canvas.
    """
    pts = np.array([[float(x), float(y)] for (x, y) in points_xy], dtype=float)
    if len(pts) < 2:
        return lo
    diag = float(np.hypot(pts[:, 0].max() - pts[:, 0].min(), pts[:, 1].max() - pts[:, 1].min()))
    err = ratio * diag
    return float(min(max(err, lo), hi))


def _fit_curve_closed(core: "np.ndarray", max_error: float, n_split: int = 4):
    """Fits a CLOSED loop (core[0] is meant to reconnect to core[-1]) by
    pre-splitting it into a handful of open arcs and fitting each one
    separately, instead of handing the whole loop to _fit_cubic as one
    piece. A single full-loop fit is degenerate for Schneider's algorithm:
    the start/end points are the same point while the chord between them is
    the entire circumference, so the least-squares tangent-scale solve
    (alpha_l/alpha_r) can become ill-conditioned and place control points
    far outside the drawing. Splitting first means every piece _fit_cubic
    ever sees has a genuine, non-degenerate chord."""
    m = len(core)
    n_split = max(2, min(n_split, m // 3)) if m >= 6 else 1
    idxs = [int(round(i * m / n_split)) for i in range(n_split)]
    out = []
    for k in range(n_split):
        i0 = idxs[k]
        if k < n_split - 1:
            i1 = idxs[k + 1]
            seg_pts = core[i0:i1 + 1]
        else:
            seg_pts = np.vstack([core[i0:], core[0:1]])  # wrap back to the start
        if len(seg_pts) < 2:
            continue
        t_hat1 = _unit_tangent(seg_pts[0], seg_pts[1])
        t_hat2 = _unit_tangent(seg_pts[-1], seg_pts[-2])
        _fit_cubic(seg_pts, t_hat1, t_hat2, max_error, out)
    return out


def fit_curve(points_xy, max_error: float = 4.0):
    """
    points_xy : liste de (x, y) capturee sur le canvas (deja simplifiee/filtree).
    max_error : erreur quadratique max toleree (en pixels^2 approx).
    Retourne une liste de segments (P0, C1, C2, P3) avec P0,C1,C2,P3 = tuples (x,y).
    """
    pts = np.array([[float(x), float(y)] for (x, y) in points_xy], dtype=float)
    dedup = [pts[0]]
    for p in pts[1:]:
        if np.linalg.norm(p - dedup[-1]) > 1e-6:
            dedup.append(p)
    pts = np.array(dedup)
    if len(pts) < 2:
        p = tuple(pts[0]) if len(pts) else (0.0, 0.0)
        return [(p, p, p, p)]

    # A CLOSED stroke is passed in here with its first and last points
    # coincident (the caller duplicates the start point to close the loop).
    # That's the degenerate case _fit_curve_closed exists for.
    if len(pts) >= 6 and np.linalg.norm(pts[0] - pts[-1]) < 1e-3:
        out = _fit_curve_closed(pts[:-1], max_error)
        return [tuple(tuple(map(float, pt)) for pt in seg) for seg in out]

    t_hat1 = _unit_tangent(pts[0], pts[1])
    t_hat2 = _unit_tangent(pts[-1], pts[-2])

    out = []
    _fit_cubic(pts, t_hat1, t_hat2, max_error, out)
    return [tuple(tuple(map(float, pt)) for pt in seg) for seg in out]

File: docs/js/curvefit.js
/*
 * curvefit.js
 * Direct JS port of the desktop app's curvefit.py: Schneider's algorithm
 * for fitting a sequence of digitized points with piecewise cubic Bezier
 * curves (Graphics Gems, 1990), including the v4.15 fix for closed loops
 * (pre-splitting into open arcs so the degenerate p0==p3 case never causes
 * a wild, out-of-bounds control point).
 */

function vsub(a, b) { return [a[0] - b[0], a[1] - b[1]]; }
function vadd(a, b) { return [a[0] + b[0], a[1] + b[1]]; }
function vscale(a, s) { return [a[0] * s, a[1] * s]; }
function vdot(a, b) { return a[0] * b[0] + a[1] * b[1]; }
function vlen(a) { return Math.hypot(a[0], a[1]); }
function vunit(a) {
  const n = vlen(a);
  return n > 1e-9 ? [a[0] / n, a[1] / n] : [0, 0];
}

function chordLengthParameterize(points) {
  const u = [0];
  for (let i = 1; i < points.length; i++) {
    u.push(u[u.length - 1] + vlen(vsub(points[i], points[i - 1])));
  }
  const total = u[u.length - 1] > 1e-9 ? u[u.length - 1] : 1.0;
  return u.map((ui) => ui / total);
}

function bezierPoint(t, ctrl) {
  const mt = 1 - t;
  return vadd(
    vadd(vscale(ctrl[0], mt ** 3), vscale(ctrl[1], 3 * mt ** 2 * t)),
    vadd(vscale(ctrl[2], 3 * mt * t ** 2), vscale(ctrl[3], t ** 3))
  );
}

function bezierDeriv(t, ctrl) {
  const mt = 1 - t;
  return vadd(
    vadd(vscale(vsub(ctrl[1], ctrl[0]), 3 * mt ** 2), vscale(vsub(ctrl[2], ctrl[1]), 6 * mt * t)),
    vscale(vsub(ctrl[3], ctrl[2]), 3 * t ** 2)
  );
}

function newtonRaphsonRootFind(ctrl, point, u) {
  const d = vsub(bezierPoint(u, ctrl), point);
  const qp = bezierDeriv(u, ctrl);
  const qpp = vadd(
    vscale(vadd(vsub(ctrl[2], vscale(ctrl[1], 2)), ctrl[0]), 6 * (1 - u)),
    vscale(vadd(vsub(ctrl[3], vscale(ctrl[2], 2)), ctrl[1]), 6 * u)
  );
  const denom = vdot(qp, qp) + vdot(d, qpp);
  if (Math.abs(denom) < 1e-9) return u;
  return u - vdot(d, qp) / denom;
}

function reparameterize(points, u, ctrl) {
  return points.map((p, i) => newtonRaphsonRootFind(ctrl, p, u[i]));
}

function generateBezier(points, u, tHat1, tHat2) {
  const n = points.length;
  const A = [];
  for (let i = 0; i < n; i++) {
    const t = u[i];
    A.push([vscale(tHat1, 3 * (1 - t) ** 2 * t), vscale(tHat2, 3 * (1 - t) * t ** 2)]);
  }
  const C = [[0, 0], [0, 0]];
  const X = [0, 0];
  const p0 = points[0];
  const p3 = points[n - 1];
  for (let i = 0; i < n; i++) {
    const t = u[i];
    C[0][0] += vdot(A[i][0], A[i][0]);
    C[0][1] += vdot(A[i][0], A[i][1]);
    C[1][0] = C[0][1];
    C[1][1] += vdot(A[i][1], A[i][1]);
    const base = vsub(bezierPoint(t, [p0, p0, p3, p3]), points[i]);
    X[0] -= vdot(A[i][0], base);
    X[1] -= vdot(A[i][1], base);
  }
  const detC0C1 = C[0][0] * C[1][1] - C[1][0] * C[0][1];
  let alphaL = 0;
  let alphaR = 0;
  if (Math.abs(detC0C1) > 1e-12) {
    const detC0X = C[0][0] * X[1] - C[1][0] * X[0];
    const detXC1 = X[0] * C[1][1] - X[1] * C[0][1];
    alphaL = detXC1 / detC0C1;
    alphaR = detC0X / detC0C1;
  }

  const segLen = vlen(vsub(p0, p3));
  let arcLen = 0;
  for (let i = 1; i < n; i++) arcLen += vlen(vsub(points[i], points[i - 1]));
  let xs = points.map((p) => p[0]);
  let ys = points.map((p) => p[1]);
  const diag = Math.hypot(Math.max(...xs) - Math.min(...xs), Math.max(...ys) - Math.min(...ys));
  const refLen = segLen > 1e-6 ? segLen : (arcLen > 1e-6 ? arcLen / 3 : 1.0);
  const fallback = refLen / 3;
  const maxAlpha = Math.max(diag, refLen, 1e-6);
  const eps = refLen > 0 ? 1e-6 * refLen : 1e-6;
  if (alphaL < eps || alphaR < eps || alphaL > maxAlpha || alphaR > maxAlpha) {
    alphaL = fallback;
    alphaR = fallback;
  }
  const c1 = vadd(p0, vscale(tHat1, alphaL));
  const c2 = vadd(p3, vscale(tHat2, alphaR));
  return [p0, c1, c2, p3];
}

function computeMaxError(points, u, ctrl) {
  let maxDist = 0;
  let splitPoint = Math.floor(points.length / 2);
  for (let i = 0; i < points.length; i++) {
    const d = vsub(bezierPoint(u[i], ctrl), points[i]);
    const dist = vdot(d, d);
    if (dist > maxDist) {
      maxDist = dist;
      splitPoint = i;
    }
  }
  return [maxDist, splitPoint];
}

function fitCubic(points, tHat1, tHat2, error, out) {
  if (points.length === 2) {
    const dist = vlen(vsub(points[0], points[1])) / 3.0;
    const p0 = points[0];
    const p3 = points[1];
    out.push([p0, vadd(p0, vscale(tHat1, dist)), vadd(p3, vscale(tHat2, dist)), p3]);
    return;
  }
  let u = chordLengthParameterize(points);
  let ctrl = generateBezier(points, u, tHat1, tHat2);
  let [maxErr, splitPoint] = computeMaxError(points, u, ctrl);
  if (maxErr < error) {
    out.push(ctrl);
    return;
  }
  for (let iter = 0; iter < 4; iter++) {
    u = reparameterize(points, u, ctrl);
    ctrl = generateBezier(points, u, tHat1, tHat2);
    [maxErr, splitPoint] = computeMaxError(points, u, ctrl);
    if (maxErr < error) {
      out.push(ctrl);
      return;
    }
  }
  if (splitPoint < 2) splitPoint = 2;
  if (splitPoint > points.length - 2) splitPoint = points.length - 2;
  let tHatCenter = vsub(points[splitPoint - 1], points[splitPoint + 1]);
  const norm = vlen(tHatCenter);
  tHatCenter = norm > 1e-9 ? vscale(tHatCenter, 1 / norm) : [0, 0];
  fitCubic(points.slice(0, splitPoint + 1), tHat1, tHatCenter, error, out);
  fitCubic(points.slice(splitPoint), vscale(tHatCenter, -1), tHat2, error, out);
}

function fitCurveClosed(core, maxError, nSplit = 4) {
  const m = core.length;
  nSplit = m >= 6 ? Math.max(2, Math.min(nSplit, Math.floor(m / 3))) : 1;
  const idxs = [];
  for (let i = 0; i < nSplit; i++) idxs.push(Math.round((i * m) / nSplit));
  const out = [];
  for (let k = 0; k < nSplit; k++) {
    const i0 = idxs[k];
    let segPts;
    if (k < nSplit - 1) {
      segPts = core.slice(i0, idxs[k + 1] + 1);
    } else {
      segPts = core.slice(i0).concat([core[0]]);
    }
    if (segPts.length < 2) continue;
    const tHat1 = vunit(vsub(segPts[1], segPts[0]));
    const tHat2 = vunit(vsub(segPts[segPts.length - 2], segPts[segPts.length - 1]));
    fitCubic(segPts, tHat1, tHat2, maxError, out);
  }
  return out;
}

/** points: array of [x,y]; returns array of [P0,C1,C2,P3] segments. */
function fitCurve(pointsXY, maxError = 4.0) {
  const dedup = [pointsXY[0]];
  for (let i = 1; i < pointsXY.length; i++) {
    if (vlen(vsub(pointsXY[i], dedup[dedup.length - 1])) > 1e-6) dedup.push(pointsXY[i]);
  }
  const pts = dedup;
  if (pts.length < 2) {
    const p = pts.length ? pts[0] : [0, 0];
    return [[p, p, p, p]];
  }
  if (pts.length >= 6 && vlen(vsub(pts[0], pts[pts.length - 1])) < 1e-3) {
    return fitCurveClosed(pts.slice(0, -1), maxError);
  }
  const tHat1 = vunit(vsub(pts[1], pts[0]));
  const tHat2 = vunit(vsub(pts[pts.length - 2], pts[pts.length - 1]));
  const out = [];
  fitCubic(pts, tHat1, tHat2, maxError, out);
  return out;
}

function adaptiveMaxError(pointsXY, ratio = 0.012, lo = 1.2, hi = 22.0) {
  if (pointsXY.length < 2) return lo;
  const xs = pointsXY.map((p) => p[0]);
  const ys = pointsXY.map((p) => p[1]);
  const diag = Math.hypot(Math.max(...xs) - Math.min(...xs), Math.max(...ys) - Math.min(...ys));
  const err = ratio * diag;
  return Math.min(Math.max(err, lo), hi);
}

function segmentsSane(segments, pts) {
  const xs = pts.map((p) => p[0]);
  const ys = pts.map((p) => p[1]);
  const x0 = Math.min(...xs), x1 = Math.max(...xs);
  const y0 = Math.min(...ys), y1 = Math.max(...ys);
  const w = Math.max(x1 - x0, 1.0), h = Math.max(y1 - y0, 1.0);
  const mx = 0.2 * w, my = 0.2 * h;
  const loX = x0 - mx, hiX = x1 + mx, loY = y0 - my, hiY = y1 + my;
  for (const seg of segments) {
    for (const [px, py] of seg) {
      if (!(px >= loX && px <= hiX && py >= loY && py <= hiY)) return false;
    }
  }
  return true;
}

function polylineAsBezier(pts) {
  const segs = [];
  for (let i = 0; i < pts.length - 1; i++) {
    const p0 = pts[i], p3 = pts[i + 1];
    const c1 = [p0[0] + (p3[0] - p0[0]) / 3, p0[1] + (p3[1] - p0[1]) / 3];
    const c2 = [p0[0] + (p3[0] - p0[0]) * 2 / 3, p0[1] + (p3[1] - p0[1]) * 2 / 3];
    segs.push([p0, c1, c2, p3]);
  }
  return segs;
}

/** fitCurve with a fallback ladder + guaranteed-in-bounds polyline fallback. */
function fitCurveSafe(pts, err) {
  for (const candidateErr of [err, Math.max(err * 4, 6.0), 20.0, 60.0]) {
    const segments = fitCurve(pts, candidateErr);
    if (segments.length && segmentsSane(segments, pts)) return segments;
  }
  return polylineAsBezier(pts);
}
